openjdk/mmtkBarrierSetC2.cpp

```diff
@@ -229,8 +238,15 @@ void MMTkBarrierSetC2::expand_allocate(PhaseMacroExpand* x,
     Node *old_eden_top = new LoadPNode(ctrl, contended_phi_rawmem, eden_top_adr, TypeRawPtr::BOTTOM, TypeRawPtr::BOTTOM, MemNode::unordered);
 
     x->transform_later(old_eden_top);
+
+    Node *offset = ConLNode::make(extra_header);
+    x->transform_later(offset);
+
+    Node *old_eden_top_plus_offset =  new AddPNode(x->top(), old_eden_top, offset);
+    x->transform_later(old_eden_top_plus_offset);
```

Comment:
This code generates the add instruction for all the allocators, which is unnecessary and may have a performance impact. the extra header should only be applied to the mark compact allocator. I am not sure if this is related with the bug though.

Reply:
Yes, but offset is a const node. I presume the impact is negligible. The bug must be related to the offset since for all other allocators with 0 offset, hotspot will not crash.

Reply:
It will depend on whether the C2 compiler does any optimization after this pass. I once had a `if` based on a const node in this code, and it turned out that the compiler does not optimize that `if`. So even though the offset is a const node, the compiler may actually generates an add instruction.